Add Contribution Guidelines for Substrate Documentation (#244)
substrate-developer-hub/substrate-developer-hub.github.io: CONTRIBUTING.md

to implement or exercise APIs or understand the internal dynamics of the blockchain system. These
standards are not intended for end-user product documentation.

### What should I document?

Document protocols, introduce essential concepts, explain how everything fits together.

- Conventions: e.g. this document about documentation, code style

- System Design: e.g. network stack, consensus, runtime, assumptions

- APIs: e.g. JSON-RPC, runtime functions

- Protocols: e.g. schemas, encodings, configuration files

- Tools: e.g. `subkey`, `build-spec`, `purge-chain`

- Workflows: e.g. environment set up, test methodologies, where to find various
  parts, how to get work done

### Where should I put documents?

Documentation that is only for developers working on creating or maintaining
a specific part of the code should be kept in the same directory as the source code.

Substrate documentation is split into three projects:

* Substrate Recipes - `substrate.dev/recipes/`

   General best practice guidelines on how to develop with Substrate.
   If you create best practice documentation about using a specific
   feature of Substrate, you should create the documentation in the same

>    directory as the other documentation for that specific feature.

Reviewer: ```suggestion
   directory as the other documentation for that specific feature. See the [contribution guidelines](https://github.com/substrate-developer-hub/recipes/blob/master/CONTRIBUTING.md) for more details.
```

Reviewer: I plan to update the CONTRIBUTING guidelines to align with this part of the fuchsia docs

- [Conceptual](#conceptual-documentation) - Understanding-oriented
  documentation

- [Procedural](#procedural-documentation)

    - How to - Goal-oriented documentation

    - Tutorial - Learning-oriented documentation

See [Documentation Types](#documentation-types) for more information.

However, comments in your code are very important for maintainability and helping other people
understand your code. See the [Code Comment Guidelines](#documentation-comments) for style guidelines
related to comments for your code.

### What documentation style guidelines should I follow?

It is important to follow documentation style guidelines to ensure that the documentation
created by a large number of contributors can flow together. See
[Documentation Style Guide](#documentation-style).

### How can I link to other docs in my documentation?

Use absolute paths starting with `/` and ending with `.md`, like [`/getting-started/using-the-substrate-scripts.md`](/substrate/node-template/runtime/src/template.rs).
Never use relative paths with ".." that point to content outside of `/docs`.

### How can I link to source code in my documentation?

> Use the `/rustdocs` to link to specific parts of the Substrate source code, for example a module's `trait`. When referencing specific lines of code, copy those lines into the documentation. However, use this sparingly.

Reviewer: I think it's worth establishing here that code shouldn't be referenced by line number. The exception might be comments to discuss docs, but maybe no docs?

Tips for leaving breadcrumbs:

- Table of contents: Add links to documentation in the left side navigation
  on substrate.dev.

- Top-down linkage: Add links from more general documents to more specific documents to help
  readers learn more about specific topics. The `conceptual` section is a good
  starting point for top-down linkage.

- Bottom-up linkage: Add links from more specific documents to more general documents to help
  readers understand the full context of the topics being discussed. Adding links from
  module, class, or protocol documentation comments to conceptual documentation overviews can be
  particularly effective.

- Sideways linkage: Add links to documents on subjects that help readers better understand the
  content of your document.

---

## Documentation Types
Documentation is an important part of any product or feature because it lets users know how to
properly use a feature that has been implemented. These guidelines are meant to be a quick and easy
reference for types of documentation. For information on documentation style guidelines, see
[Documentation Style Guide](#documentation-style).

> ### Conceptual, procedural, or reference documentation

Reviewer: Maybe this is already in here somewhere, but it should be clear where each of these categories of documentation go. **We need to make the mapping from these categories to the places where we store/create documentation explicit**.  This is pretty subjective but I just feel like this is too easily ignored otherwise. 

Maybe new docs might adhere to a template where the author specifies (1) where they intend to put the docs (2) what categories of docs are covered in their new doc (3) how their new doc communicates/references the other categories of docs. To be clear, I don't like excessive templates or adding unnecessary friction, but I don't want us to ignore the step where we first categorize the new material before finishing a writeup because it should clearly fall into one of these categories and our current workflow isn't oriented around these categories.

Author: I don't think process this needs to live in our contribution guidelines. We could create a PR template which does something like this.

Internally, I am a fan of defining ahead of time what kind of doc it is, where it goes, etc...

including, but not limited to APIs and CLIs. Reference documentation should allow the user to
understand how to use a specific feature quickly and easily.

**You should write a conceptual document** if you plan on explaining a concept about a product.
Conceptual documents explain a specific concept, but for the most part they do not include actual
examples. They provide essential facts, background, and diagrams to help your readers build a
foundational understanding of a product or topic. You should not explain industry standards that
your audience should be familiar with, for example, TCP/IP. You might explain how this concept ties
in with your feature, but you should not explain the basics behind that industry-standard concept.

**You should write a procedural document** if you plan on explaining to a user how to use a specific
feature and are able to guide a user through simple numbered steps. Procedural documents tend to
reinforce the concepts that were explained in a conceptual document by giving one or more
examples that might be useful for users.

Procedural documents are divided into two categories:

> - **How-to** - Consider writing a how-to when you want to help the user accomplish a very specific

Reviewer: could I just add recipes in here somehow? like I said further above, I think the mapping from these categorizations to our existing workflow should be more explicit

something like "If the how-to pertains to runtime/module development patterns, it probably belongs in the recipes." Idk it feels out of place because there is no clear references to our doc workflow in this section but I'm saying maybe there should be.

Author: I think we should keep this content free of the implementation details.  The Substrate Recipes is a collection of "how-to"s

  goal.

- **Tutorial** - Consider writing a tutorial when you want to help the user learn about a bigger
  goal that might involve working with multiple parts of a product or feature. The tutorial should not

>   go over 60 minutes and should provide the user with a specific result.

Reviewer: workshops are a different category?

Author: See below

Reviewer: So, if workshops are tutorials then we won't have any workshops that go over 60 minutes? cc @riusricardo

Reviewer: if workshops are tutorials as mentioned further below,  are all in-person workshops limited to 60 minutes? cc @riusricardo

a feature specific `images/` directory or a common `images/` directory. You should also save
the source file of your images in an `images/src/` directory.

Good conceptual documentation usually includes:

- **Description** rather than instruction

- **Background** concepts

- **Diagrams** or other visual aids (preferably in .png format)

- **Links** to how-to and/or reference docs
After writing your document, it is good practice to proofread the document, put yourself in the
user's shoes (no longer being the expert that developed the feature), and try to answer these
questions:

- Does the information in the document explain the concept completely?

- Is there information that is not needed for this concept? If so, remove it.

- Is there unnecessary detail about how things might work in the background?

- If I am the user, is there additional information I would have liked to know?

Then, add your feedback into your document.

### Procedural documentation

Procedural documents are divided into two categories:

- **How-to** - Consider writing a how-to when you want to help the user accomplish a very specific
  goal.

> - **Tutorial** - Consider writing a tutorial when you want to help the user learn about a bigger goal

Reviewer: workshop?

Author: I would be a fan of changing the use of "workshop" to "tutorial".

To me, a tutorial is a guide for doing something, a workshop is doing a tutorial live with people.

"We write tutorials, and present them at workshops"

Note the reason why it was called the `substrate-collectables-workshop` was due to the fact it was originally written for a particular workshop. It just kind of exploded from there.

Reviewer 2: I agree with Shawn, during a workshop you will follow a tutorial. We should refer to them as tutorials on the devhub.

- **Use active voice.** You should always try to write in the active voice since passive voice can
  make sentences very ambiguous and hard to understand. There are very few cases where you should
  use the passive voice for technical documentation.

  - Active voice - the subject performs the action denoted by the verb.

    - "The operating system runs a process." This sentence answers the question on what is
      happening and who/what is performing the action.

  - Passive voice - the subject is no longer _active_, but is, instead, being acted upon by the
    verb - or passive.

    - "A process is being run." This sentence is unclear about who or what is running the process.
      You might consider "a process is run by the operating system", but the object of the action
      is still made into the subject of the sentence, which indicates passive voice. Passive voice
      tends to be wordier than active voice, which can make your sentence unclear. In most cases,
      if you use "by" this indicates that your sentence might be still be in passive voice. A
      correct way of writing this example is, "The operating systems runs the process."

> - **Do not list future plans for a product/feature.** "In the future, the product will have no

Reviewer: This will seem hard to adhere to for us.

Author: This seems important for us to enforce within our world of documentation. I think we have been very bad in the past talking about features like off-chain workers, which is _still_ not ready.

I am a big fan of this rule. Anything speculative can live with the code or in issues.

Reviewer: Who says when a feature is ready? Where do we draw the line on what is speculative vs non-speculative? I'm just saying that these things aren't so clear. We're working with speculative technology and documenting speculative code.

Also, I think raising awareness of offchain workers before was useful even if it isn't completely finished. Because of this conceptual documentation, some developers are positioning themselves to take advantage of this feature when it is ready and that's what we want.

Reviewer 2: Isn't versioning supposed to solve part of this issue? A feature could be tested on a specific Substrate version.
For the second part, I'll also prefer to be less explicit about future features until we explore and play with them.

Reviewer: So, all docs are limited to code that is feature complete now? We're going to keep getting a better understanding of what a good runtime architecture will look like and I don't feel comfortable keeping this to ourselves until we get around to writing usage-oriented tutorials. 

For example, fixed point arithmetic is still undergoing refactoring. The storage api for child tries isn't completely finished nor are all parts of off-chain workers. Swappable consensus will not be ready for some time. Also, we keep claiming that substrate chains can trivially be turned into parachains but cumulus is also still a wip.

So, yes, maybe we are less loud about these features but we should continue communicating our expectations of future best practices in conceptual documentation, even if we don't have immediate procedural examples. *I'm mostly just being pedantic because I don't think this can be enforced consistently, but it's fine to discuss later on a case by case basis.*

    Does "this" refer to fast, efficient, or operating system? Consider using: "The operating system
    is well designed because it is fast and efficient."

- **Keep sentences fairly short and concrete.** Using punctuation allows your reader to follow
  instructions or concepts. If by the time you read the last word of your sentence, you can't
  remember how the sentence started, it is probably too long. Also, short sentences are much easier
  to translate correctly.

- **Know your audience.** It is good practice to know your audience before you write documentation.
  Your audience can be, for example, developers, end-users, integrators, etc. and they can have varying
  degrees of expertise and knowledge about a specific topic. Knowing your audience allows you to
  understand what information your audience should be familiar with. When a document is meant for a
  more advanced audience, it is best practice to state it up front and let the user know
  prerequisites before reading your document.

- **Use markdown.** You must create documentation in markdown (.md) and keep the markdown file
  wrapped to a 120 character column size.

> - **Be respectful**

Reviewer: ```suggestion
- **Be respectful** -- adhere to the [Rust CoC](https://www.rust-lang.org/policies/code-of-conduct)
```

### What should I document?

Document protocols, introduce essential concepts, explain how everything fits together.

- Conventions: e.g. this document about documentation, code style

- System Design: e.g. network stack, consensus, runtime, assumptions

- APIs: e.g. JSON-RPC, runtime functions

- Protocols: e.g. schemas, encodings, configuration files

- Tools: e.g. `subkey`, `build-spec`, `purge-chain`

- Workflows: e.g. environment set up, test methodologies, where to find various
  parts, how to get work done

### Where should I put documents?

Documentation that is only for developers working on creating or maintaining
a specific part of the code should be kept in the same directory as the source code.

Substrate documentation is split into three projects:

* Substrate Recipes - `substrate.dev/recipes/`

   General best practice guidelines on how to develop with Substrate.
   If you create best practice documentation about about using a specific
   feature of Substrate, you should create the documentation in the same
   directory as the other documentation for that specific feature.

* Substrate Rust Documentation - `substrate.dev/rustdocs/`

>    Auto-generated Rust documentation from the Substrate repository. See [Documentation Comments](#documentation-comments)

Reviewer 3: ```suggestion
   Auto-generated Rust documentation from the Substrate repository. See [Documentation Comments](#documentation-comments).
```

> * Substrate Developer Documentation - `substrate.dev/docs/`

Reviewer 4: What about substrate.dev/tutorials?

Author: Its more of an index no?

The actual tutorials, if in the substrate.dev site, live at `https://substrate.dev/docs/tutorials/*/`

But we know these tutorials also live in other places (something we might want to change?)

Maybe we could consider creating a `codelab` like platform where we place all of our tutorials outside of docusaurus.

Reviewer 2: We should try to keep this path  `substrate.dev/tutorials` to make it easy for a user to remember and get back to a tutorial.

Reviewer: imo it's worth linking to [tutorials](https://substrate.dev/tutorials/) until some `codelab` initiative is further along. 

The links for each tutorial might point to different sites/repos, but I don't think that's a big deal. The point of the tutorials tab on substrate.dev is to provide a central place to keep all of these `codelab`-esque examples.

Author: Sorry, so what is the consensus here?

I think we agree that `substrate.dev/tutorials/` is not a place "Where should I put documents" (the title of this section).

It acts as an index for where the documents actually live.

@JoshOrndorff if you agree you can resolve this

Reviewer 4: Agreed that tutorials is an index and should be kept in place. Agreed that it's okay to have some tutorial be external links at least in the medium term possibly indefinitely. Agreed that tutorials that live as part of devhub should go in `/docs`.

I think it just feels funny that we're referring to those paths as domain names rather than file paths in the repo. That's not a big deal for me though.

   In the `/docs/` directory, you should create your
   documentation or images in one of these sub-directories:
    
   *  `development`
        Instructions, tutorials, and procedural documentation for developers

>         that are working on Substrate. This directory includes documentation

Reviewer 3: ```suggestion
        who are working on Substrate. This directory includes documentation
```

- [Conceptual](#conceptual-documentation) - Understanding-oriented
  documentation

- [Procedural](#procedural-documentation)

    - How to - Goal-oriented documentation

    - Tutorial - Learning-oriented documentation

See [Documentation Types](#documentation-types) for more information.

However, comments in your code are very important for maintainability and helping other people
understand your code. See the [Code Comment Guidelines](#documentation-comments) for style guidelines
related to comments for your code.

### What documentation style guidelines should I follow?

It is important to try to follow documentation style guidelines to ensure that the documentation
created by a large number of contributors can flow together. See
[Documentation Style Guide](#documentation-style).

### How can I link to other docs in my documentation?

Use absolute paths starting with `/` and ending with `.md`, like [`/getting-started/using-the-substrate-scripts.md`](/substrate/node-template/runtime/src/template.rs).
Never use relative paths with ".." that point to content outside of `/docs`.

### How can I link to source code in my documentation?

> Use the `/rustdocs` to link to specific parts of the Substrate source code, for example `traits` of a module. When referencing specific lines of code, copy those lines into the documentation, however, use this sparingly.

Reviewer 3: ```suggestion
Use the `/rustdocs` to link to specific parts of the Substrate source code, for example a module's `trait`. When referencing specific lines of code, copy those lines into the documentation. However, use this sparingly.
```

### How can I expose my documentation?

Documentation is only useful when users can find it. Adding links to or from existing documentation
greatly improves the chances that someone can find your documentation.

Tips for leaving breadcrumbs:

> - Table of contents: Add links to documentation in the left sided navigation

Reviewer 3: ```suggestion
- Table of contents: Add links to documentation in the left side navigation
```

  as the Substrate runtime.

- [Procedural](#procedural-documentation)

    - How-to - Documentation that provides steps on how to accomplish a goal such as setting up your computer to compile Substrate.

    - Tutorial - Documentation that provides steps of a learning path (this tends to be a much bigger
      procedure than a how-to) such as create a Substrate runtime module.

**You should write a reference document** if you need to provide information about parts of a system
including, but not limited to APIs and CLIs. Reference documentation should allow the user to
understand how to use a specific feature quickly and easily.

**You should write a conceptual document** if you plan on explaining a concept about a product.
Conceptual documents explain a specific concept, but for the most part they do not include actual
examples. They provide essential facts, background, and diagrams to help your readers build a
foundational understanding of a product or topic. You should not explain industry standards that
your audience should be familiar with, for example, TCP/IP. You might explain how this concept ties

> in with your feature, but you should not explain the basics behind that industry standard concept.

Reviewer 3: ```suggestion
in with your feature, but you should not explain the basics behind that industry-standard concept.
```

**You should write a procedural document** if you plan on explaining to a user how to use a specific
feature and are able to guide a user through simple numbered steps. Procedural documents tend to
reinforce the concepts that were explained in a conceptual document by giving one or more
examples that might be useful for users.

Procedural documents are divided into two categories:

> - **How-to** - Consider writing a how to when you want to help the user accomplish a very specific

Reviewer 3: ```suggestion
- **How-to** - Consider writing a how-to when you want to help the user accomplish a very specific
```

- How does a transistor work? This is a conceptual document that would be geared towards a more
  advanced user.

- Functions of the car radio. This is a reference document.

- How a new technology improved the car radio. This is a conceptual document.

> Note: A feature may require more than one type of document. You may decide that your feature
requires just reference documentation or that you need reference, conceptual, and how to
documentation.

### Reference documentation

Reference documentation should provide information about parts of a system including, but not
limited to APIs and CLIs. The style of reference documentation should be the same for all reference
documentation of that type. For example, API documentation should define all of the API's parameters,
indicate if a parameter is required or optional, and show examples of the use of the API. These
examples should be very generic and simple. If you feel like you need a more elaborate example,
consider creating a procedural document to reinforce your reference documentation.

For the style guide for API documentation, see
[API style guide](../development/api/documentation.md).

> ### Conceptual documentation

Reviewer 5: Add empty line above here.

Conceptual documentation should try to be brief and for the most part should not go above 1 page.
If you need to write more than one page to describe a concept, consider breaking that concept into
sub-concepts by using headings. By keeping your document brief you achieve the following:

- You do not overwhelm your reader with a wall of text.

> - Avoid losing the reader while they read your document.

Reviewer 3: ```suggestion
- You avoid losing the reader while they read your document.
```

The first paragraph should try to be a brief summary of your document, this should allow the user to
quickly read through it, determine what the document covers, and if this is relevant to what they
want to learn. If your document has multiple headings, you should include a bulleted list with the
high-level headings after this first paragraph.
You should use graphics, images, or diagrams to reinforce certain concepts. The text that comes
before and after the graphic should explain what the graphic shows. Images should be saved in
a feature specific `images/` directory or a common `images/` directory. You should also save

> the source file of your images in a `images/src/` directory.

Reviewer 3: the?
```suggestion
the source file of your images in an `images/src/` directory.
```

Good conceptual documentation usually includes:

- **Description** rather than instruction

- **Background** concepts

- **Diagrams** or other visual aids (preferably in .png format)

- **Links** to how-to and/or reference docs
After writing your document, it is good practice to proofread the document, put yourself in the
user's shoes (no longer being the expert that developed the feature), and try to answer these
questions:

- Does the information in the document explain the concept completely?

- Is there information that is not needed for this concept? If so, remove it.

- Is there unnecessary detail about how things might work in the background?

> - If I am the user, is there additional I would have liked to know?

Reviewer 3: ```suggestion
- If I am the user, is there additional information I would have liked to know?
```

Then, add your feedback into your document.

### Procedural documentation

Procedural documents are divided into two categories:

> - **How-to** - Consider writing a how to when you want to help the user accomplish a very specific

Reviewer 3: I would hyphenate "how to" any time it's used as a compound adjective or noun.
```suggestion
- **How-to** - Consider writing a how-to when you want to help the user accomplish a very specific
```

  goal.

- **Tutorial** - Consider writing a tutorial when you want to help the user learn about a bigger goal
  that might involve working with multiple parts of a product or feature.

Procedural documentation should try to be brief and each task within your documentation should try
to avoid going above 10 steps (tutorials can be much longer, but should not exceed 45-60 minutes for
a user to complete). You should divide long procedures into multiple sub-tasks to try to keep tasks
manageable for a user. For example, if you wanted to write a procedural document for taking care of
a dog, you might have a table of content that looks like this:

> How to take care of a dog:
> 
> - Feeding a dog
> 
> - Washing a dog
> 
> - Trimming a dog's nails
> 
> - Brushing a dog
> 
> - Playing with a dog

#### Difference between a tutorial and a how to

> At a very high-level, a tutorial is essentially a large how to, composed of various smaller how tos.

Reviewer 5: ```suggestion
At a very high-level, a tutorial is essentially a large how-to, composed of various smaller how-tos.
```

Tutorials are great when you want to give the user a hand-held experience of working through a task,
especially if this task is considered a little more complicated and might involve working with

> various areas of a product. On the other hand, a how to should describe the steps on how to work

Reviewer 5: ```suggestion
various areas of a product. On the other hand, a how-to should describe the steps on how to work
```

through a minor task that should only involve a single area of a product.

Consider the following when you think that you might need to create a tutorial:

- How many tutorials are planned for this general feature? Keep in mind that you do not want a
  whole documentation set to just be tutorials, use them in moderation.

- Tutorials should be self-contained, avoid creating links to other tutorials, other how-tos or

>   other information that might have a user leave the actual tutorial. It is ok to provide links to

Reviewer 3: ```suggestion
  other information that might have a user leave the actual tutorial. It is OK to provide links to
```

  conceptual documents that can enhance a user's knowledge for a given topic.

- Would this procedural documentation benefit from having a very specific example through a
  tutorial?

- Do you want to expose an exciting feature from the product through the tutorial? This helps you
  highlight a neat feature that a user might not know about without doing a tutorial.

#### General procedural documentation guidelines

- Each task or subtask should have a paragraph that lets a user know what the task is about and
  what a user should be able to do after performing the steps.

- Use screenshots or graphics to assist a user in navigating a user interface (UI).

- A procedural document should not have to explain any concepts to a user, but should reference
  conceptual documents in case a user does not know about a certain concept. For example, a
  procedure with a reference to a conceptual document might look like this:

   > Configure the server with the appropriate configuration. For more information about server
   configurations, see "server configuration".

> - Avoid giving the users multiple paths to select when working through procedures. When you avoid

Reviewer 3: ```suggestion
- Avoid giving the user multiple paths to select from when working through procedures. When you avoid
```

Tips for writing effective documentation comments:

- Write in plain U.S. English.

- Write complete sentences and paragraphs.

- Write clear and brief comments, no more than a few sentences.

- Follow the approved [style guide for the Rust programming language](https://doc.rust-lang.org/1.0.0/style/).

- Avoid creating running words that are not compound words. For example "notready" is two words
  run together. Use an appropriate separator, for example "not ready", "notReady", "not_ready", or
  "not-ready").

- Always add value. Don't restate what is already indicated by the type signature.

- Describe units of measure and integrity constraints of variables.

- Link to conceptual documentation for more complete descriptions of how APIs fit together as a
  whole.

---

## Documentation Style

It is important to create documentation that follows similar guidelines. This allows documentation
to be clear and concise while allowing users to easily find necessary information. For information
about the complete documentation standards, see
[Documentation Standards](documentation_standards.md).

These are general style guidelines that can help create clearer documentation:

> - **Write in plain U.S. English.** You should write in plain U.S. English and try to avoid over

Reviewer 3: "over-complicated"

  complicated words when you describe something.

- **Avoid using pronouns such as "I" or "we".** These can be quite ambiguous when someone reads the
  documentation. It is better to say "You should do…." instead of "We recommend that you do….". It

>   is ok to use "you" as this allows the documentation to speak to a user.

Reviewer 3: ```suggestion
  is OK to use "you" as this allows the documentation to speak to a user.
```

- **In most cases, avoid future tense.** Words such as "will" are very ambiguous. For example "you
  will see" can lead to questions such as "when will I see this?". In 1 minute or in 20 minutes? In
  most cases, assume that when someone reads the documentation you are sitting next to them and
  reading the instructions to them.

- **Use active voice.** You should always try to write in the active voice since passive voice can
  make sentences very ambiguous and hard to understand. There are very few cases where you should
  use the passive voice for technical documentation.

  - Active voice - the subject performs the action denoted by the verb.

    - "The operating system runs a process." This sentence answers the question on what is
      happening and who/what is performing the action.

  - Passive voice - the subject is no longer _active_, but is, instead, being acted upon by the
    verb - or passive.

    - "A process is being run." This sentence is unclear about who or what is running the process.
      You might consider "a process is run by the operating system", but the object of the action

>       is still made into the subject of the sentence which indicates passive voice. Passive voice

Reviewer 3: ```suggestion
      is still made into the subject of the sentence, which indicates passive voice. Passive voice
```

      tends to be wordier than active voice, which can make your sentence unclear. In most cases,
      if you use "by" this indicates that your sentence might be still be in passive voice. A

>       correct way of writing this example is "The operating systems runs the process."

Reviewer 3: ```suggestion
      correct way of writing this example is, "The operating systems runs the process."
```

- **Do not list future plans for a product/feature.** "In the future, the product will have no
  bugs." This leads to the question as to when this would happen, but most importantly this is not
  something that anyone can guarantee will actually happen.

- **Do not talk about how certain features work behind the covers unless it is absolutely necessary.**
  Always ask yourself, "Is this text necessary to understand this concept or to get through these
  instructions?" This also leads to shorter (less maintenance) and more concise (happier readers)
  documentation.

- **Avoid using uncommon words, highly technical words, or jargon that users might not understand.**
  Also, avoid using idioms such as "that's the way the cookie crumbles", while it might make sense

>   to you, it could not translate well into another language. Keep in mind that a lot of users are

Reviewer 3: ```suggestion
  to you, it may not translate well into another language. Keep in mind that a lot of users are
```

  non-native English speakers.

- **Use compound words correctly.** Use the compound words that give the correct meaning.
  For example, "set up" (verb)  and "setup" (noun) have different meanings.

- **Avoid using words such as "best" or "great" since these are all relative terms.** How can you
  prove that "this operating system is the best?"

- **Avoid referencing proprietary information.** This can refer to any potential terminology or

>   product names that may be trademarked or any internal information (API keys, machine names, etc…)

Reviewer 3: ```suggestion
  product names that may be trademarked or any internal information (API keys, machine names, etc…).
```

- **Avoid starting a sentence with "this" since it is unclear what "this" references.**

  - For example: "The operating system is fast and efficient. This is what makes it well designed."
    Does "this" refer to fast, efficient, or operating system? Consider using: "The operating system

>     is well designed because it is fast and efficient".

Reviewer 3: ```suggestion
    is well designed because it is fast and efficient."
```

- **Keep sentences fairly short and concrete.** Using punctuation allows your reader to follow
  instructions or concepts. If by the time you read the last word of your sentence, you can't
  remember how the sentence started, it is probably too long. Also, short sentences are much easier
  to translate correctly.

- **Know your audience.** It is good practice to know your audience before you write documentation.

>   Your audience can be, for example, developers, end-users, integrators, and they can have varying

Reviewer 3: ```suggestion
  Your audience can be, for example, developers, end-users, integrators, etc. and they can have varying
```
